cf/field.py

```diff
@@ -985,80 +1004,123 @@ def _binary_operation(self, other, method):
             data_axes = f.get_data_axes()
             for axis in f.domain_axes(todict=True):
                 identity = None
-                key = None
-                coord = None
-                coord_type = None
 
-                key, coord = f.dimension_coordinate(
-                    item=True, default=(None, None), filter_by_axis=(axis,)
-                )
-                if coord is not None:
-                    # This axis of the domain has a dimension
-                    # coordinate
-                    identity = coord.identity(strict=True, default=None)
-                    if identity is None:
-                        # Dimension coordinate has no identity, but it
-                        # may have a recognised axis.
-                        for ctype in ("T", "X", "Y", "Z"):
-                            if getattr(coord, ctype, False):
-                                identity = ctype
-                                break
-
-                    if identity is None and relaxed_identities:
-                        identity = coord.identity(relaxed=True, default=None)
-                else:
-                    key, coord = f.auxiliary_coordinate(
-                        item=True,
-                        default=(None, None),
+                if self.is_discrete_axis(axis):
+                    # This is a discrete axis whose identity is
+                    # inferred from all of its auxiliary coordinates
+                    x = {}
+                    for key, aux_coord in f.auxiliary_coordinates(
                         filter_by_axis=(axis,),
-                        axis_mode="exact",
+                        axis_mode="and",
+                        todict=True,
+                    ).items():
+                        identity = aux_coord.identity(
+                            strict=True, default=None
+                        )
+                        if identity is None and relaxed_identities:
+                            identity = aux_coord.identity(
+                                relaxed=True, default=None
+                            )
+                        if identity is not None:
+                            x[identity] = key
+
+                    if x:
+                        # E.g. {2:3, 4:6, 1:7} -> (1, 2, 4), (7, 3, 6)
+                        identity, keys = tuple(zip(*sorted(x.items())))
```

Comment:
Hi Sadie - I don't think there's anything up with the API, here. The API allowed us to define `x` really easily as a dictionary of identity/key pairs. Whilst the next line looks horrible, it's just saying "I'd like to get the sorted identities (sorted so that they're comparable between fields) and their corresponding keys", using standard library tools.

I'll add a comment to this end.

Reply:
b7c1723416e453a2d38db0e788c60257b2cb1a5f

Reply:
Thanks for clarifying. It sounds to me from your explanation like I was missing the key aspect of them being sorted through those operations. So, all good. Marking as resolved.